Feat(nextjs-auth): New authentication management

SamSalvatico opened the pull request:
### Description

With the new version we don't store any Logto context data on our side anymore, we just parse data from Logto
Postgres code has been removed

**Please note**: Organization selector at the moment is not exported, will be exported in a second time
## Type

- [ ] **Dependency upgrade**
- [ ] **Bug fix**
- [x] **New feature**
- [x] **Dev change**
- [ ] **Additional tests**
- [ ] **Documentation**
- [ ] **Other**

SamSalvatico commented:
Here you can find a usage example
https://github.com/ogcio/life-events/tree/feat/new-sessions-with-ccokies/apps/messaging

alfonsograziano commented on packages/nextjs-auth/src/types.ts:
```diff
@@ -1,100 +1,79 @@
-};
-
-export type AuthConfig = LogtoNextConfig;
+export const DEFAULT_ORGANIZATION_ID = "ogcio";
```
Not-blocker: shall we get rid of this? As far as I can see, this is only used for sorting, but AFAIK, we will not actively use the `ogcio` organization in production as we will only use the organizations of each department

alfonsograziano commented on packages/nextjs-auth/src/parse-logto-context.ts:
```diff
@@ -0,0 +1,171 @@
+} from "./types.js";
+import { DEFAULT_ORGANIZATION_ID } from "./types.js";
+
+export function parseOrganizationInfo(
```
Non-blocker: given that we're aiming for a rewrite of this auth layer, it might be worth adding some docs (even basic ones) that explain what each function is doing. I'm saying this as in the old implementation sometimes we struggled to understand the details, so we might avoid this issue in the new implementation with some documentation and adding comments over specific behaviors

alfonsograziano commented on packages/nextjs-auth/src/parse-logto-context.ts:
```diff
@@ -0,0 +1,171 @@
+    organizationRoles = new Set<string>(context.claims.organization_roles);
+  }
+
+  if (context.userInfo && Array.isArray(context.userInfo.organization_roles)) {
```
Can we get the organization_roles both from userInfo and claims? Is that an inconsistence in the Logto SDK?

SamSalvatico replied:
We will get organization_roles from claims if specific custom claims are requested
We are not using this at the moment, but can help for the future

alfonsograziano replied:
Interesting, I wasn't aware of this. Thanks :D

alfonsograziano commented on packages/nextjs-auth/src/parse-logto-context.ts:
```diff
@@ -0,0 +1,171 @@
+      organizationRoles = new Set<string>();
+    }
+
+    organizationRoles = new Set<string>([
```
This to me looks like a repetition. In this function there might be some room for a refactor to make it simpler

Thurfjell commented on packages/nextjs-auth/src/index.ts:
```diff
@@ -0,0 +1,10 @@
+// export { SelectedOrganizationHandler } from "./selected-organization-handler.js";
```
```suggestion
```

Thurfjell commented on packages/nextjs-auth/src/authentication-routes.ts:
```diff
@@ -0,0 +1,21 @@
+import type { LogtoNextConfig } from "@logto/next";
+import { handleSignIn, signIn, signOut } from "@logto/next/server-actions";
+import { addInactivePublicServantScope } from "./inactive-public-servant.js";
+// import { SelectedOrganizationHandler } from "./selected-organization-handler.js";
```
```suggestion
```

alfonsograziano commented on packages/nextjs-auth/src/user-context-handler.ts:
```diff
@@ -0,0 +1,160 @@
+
+    return context.isPublicServant;
+  }
+  async isCitizen(): Promise<boolean> {
```
This is interesting - we might have some false positives here when we start to add new roles. It would be great to have a unique way to recognize whether you're a citizen. One idea might be to add (in the citizen scopes) something like `bb:citizen:*` that we can then leverage here to uniquely identify whether a user is a citizen in our system

Thurfjell commented on packages/nextjs-auth/src/inactive-public-servant.ts:
```diff
@@ -0,0 +1,23 @@
+  "inactive-ps-org:Inactive Public Servant";
+
+export function addInactivePublicServantScope(
+  config: LogtoNextConfig,
```
If you don't want to mutate config, you're gonna need to clone the array (scopes). iirc this will mutate config.scopes.

alfonsograziano commented on packages/nextjs-auth/src/user-context-handler.ts:
```diff
@@ -0,0 +1,160 @@
+  }
+  async hasPermissions(
+    permissions: string[],
+    matchMethod: "OR" | "AND" = "OR",
```
shall we get rid of this `matchMethod` here and in the `hasPermissions`? At the moment - based on the discussions we had in the past, AFAIK, we're only using the "OR" and we don't want to use the "AND" for now - it might be worth checking whether we're using it somewhere though

alfonsograziano commented on packages/nextjs-auth/src/user-context-handler.ts:
```diff
@@ -0,0 +1,160 @@
+
+    return hasPermissions(scopes, permissions, { method: matchMethod });
+  }
+  async getTokenFromContext(): Promise<string | undefined> {
```
I'm still a little bit confused about the difference between this token and the `getAccessToken` method as, here we're returning `context.accessToken`. There might be room to improve the name of the method, add some docs, or better explain how this method is supposed to be used

SamSalvatico replied:
I've added some info here
We still have to check which scopes are returned for this token, so maybe we can reuse this token instead of requesting a new one

alfonsograziano reviewed (commented):
It looks super nice overall - cannot wait to implement it. I just left a few minor comments here and there. Thanks Sam, we owe you a beer!

SamSalvatico commented on packages/nextjs-auth/src/authentication-routes.ts:
```diff
@@ -0,0 +1,21 @@
+    return signIn(addInactivePublicServantScope(config));
+  },
+  async logout(config: LogtoNextConfig, redirectUri: string): Promise<void> {
+    //SelectedOrganizationHandler.delete();
```
Thank you Ludwig, but as I wrote in the PR description I would prefer to maintain these comments because we are near to need to implement a centralized organization selector, once we are ready to do that we can just uncomment those entries

Gonna check the other comments!

SamSalvatico commented on packages/nextjs-auth/src/user-context-handler.ts:
```diff
@@ -0,0 +1,168 @@
+  getLogtoContext,
+  getOrganizationToken,
+} from "@logto/next/server-actions";
+import { redirect } from "next/navigation.js";
```
When we will need to move to 15 we will need to update logto/next sdks too

Then, let's say that the nextjs-auth (this package) version 4 is using next 14
nextjs-auth version 5 will use next 15, then we should be fine based on versioning

SamSalvatico commented:
I should have addressed all the comments (but for the isCitizen one, already discussed with Alfonso, we will find a way in a future release), so please give a look

I've already briefly tested locally, once you are okay with this we can try to implement in one service (messaging?) and run some tests thanks to David

Thurfjell reviewed (approved):
🌟

alfonsograziano reviewed (approved):
LGTM -Thanks Sam!
